Add tests with RelStorage and sqlite.

jamadden opened the pull request:
Fixes #30.

- Refactor the tests to use a mixin class that defines the methods to create and cleanup storages.
- Use unittest.skip decotarors instead of manually listing out the tests to run in test_suite(). This is simpler now that there are more test classes.
- Add tests for history-free and history-preserving RelStorage storages. History preserving is tested locally, but will be skipped on CI because I haven't released RelStorage 3.3 yet.
- Fix a test hygiene issue that caused subsequent tests to fail after test_skipped_types_are_left_untouched was run.

jamadden committed:
Add tests with RelStorage and sqlite.

Fixes #30.

- Refactor the tests to use a mixin class that defines the methods to create and cleanup storages.
- Use unittest.skip decotarors instead of manually listing out the tests to run in test_suite(). This is simpler now that there are more test classes.
- Add tests for history-free and history-preserving RelStorage storages. History preserving is tested locally, but will be skipped on CI because I haven't released RelStorage 3.3 yet.
- Fix a test hygiene issue that caused subsequent tests to fail after test_skipped_types_are_left_untouched was run.

jamadden committed:
RelStorage+sqlite on Python 3 requires 3.6+

mgedmin commented on src/zodbupdate/tests.py:
```diff
@@ -809,17 +835,17 @@ def test_encode_binary_leaves_none_untouched(self):
-        self.assertEquals(result, False)
-        self.assertEquals(mock['foo'], None)
+        self.assertEqual(result, False)
+        self.assertEqual(mock['foo'], None)
```
Minor: might as well use `assertFalse` and `assertIsNone`.

jamadden replied:
Will do. I replaced `assertEquals` originally only because it was showing deprecation warnings.

mgedmin commented on src/zodbupdate/tests.py:
```diff
@@ -809,17 +835,17 @@ def test_encode_binary_leaves_none_untouched(self):
-        self.assertEquals(result, False)
-        self.assertEquals(mock['foo'], None)
+        self.assertEqual(result, False)
+        self.assertEqual(mock['foo'], None)
```
likewise

mgedmin commented on src/zodbupdate/tests.py:
```diff
@@ -1266,11 +1293,76 @@ def test_loaded_renames_override_missing_interfaces(self):
+                   data-dir %s
+                </sqlite3>
+            </relstorage>
+            """ % (os.path.join(self.tmpblob, 'blobs'), self.tmpblob,)
```
Minor: I'd've used "%import ... blob-dir {blobdir} ... ".format(blobdir=...) to avoid having to double the percents.

mgedmin commented on src/zodbupdate/tests.py:
```diff
@@ -1266,11 +1293,76 @@ def test_loaded_renames_override_missing_interfaces(self):
+            """ % (os.path.join(self.tmpblob, 'blobs'), self.tmpblob,)
+        )
+
+    _tearDownStorage = lambda self: None
```
This is a bit surprising to me.

If _tearDownStorage is meant to remove all the data files, then this one doesn't do that.

If _tearDownStorage is meant to remove any additional data files that live outside the temporary directory that gets shutils.rmtree'd, then why does the base class take pains to remove the file storage files manually?  It could be a `pass` and relstorage wouldn't need to override it.

jamadden replied:
Yeah, that's poorly factored. In the effort to keep changes minimal, I left an unclear division of responsibility. I'll clean that up. 

Thanks for calling that out!

jamadden replied:
> If _tearDownStorage is meant to remove any additional data files that live outside the temporary directory that gets shutils.rmtree'd, then why does the base class take pains to remove the file storage files manually?

The underlying reason is that originally, the FileStorage data file and its blobs directory were completely different parts of the filesystem tree. Only the blobs directory was `shutils.rmtree`d. I changed that to put both the data file and blobs directory under the temporary directory (making FileStorage lay things out like RelStorage was).

mgedmin replied:
Oh, oops, and I didn't even notice!

I like your changes a lot!

mgedmin reviewed (approved):
LGTM

jamadden committed:
Refactor tests to more fully separate storage creation/teardown concerns.

Make FileStorage its own mixin, and put everything in the single temp directory, which is automatically torn down.

mgedmin commented on src/zodbupdate/tests.py:
```diff
@@ -140,8 +143,7 @@ class OtherFactory(persistent.Persistent):
 
-        self.tmphnd, self.dbfile = tempfile.mkstemp()
-        self.tmpblob = tempfile.mkdtemp()
+        self.temp_dir = tempfile.mkdtemp('.zodbupdate')
```
:100:

mgedmin reviewed (approved):
LGTM

jamadden commented:
Thanks!
